lib/utils/format.ts

```diff
@@ -0,0 +1,256 @@
+ * @param Model
+ */
+export function mapOptionFieldNames<M extends Model>(
+  options: FinderOptions<M['_attributes']>,
+  Model: ModelStatic<Model>,
+): MappedFinderOptions<M['_attributes']> {
+  const out: MappedFinderOptions<M['_attributes']> = options;
+
+  if (Array.isArray(options.attributes)) {
+    out.attributes = options.attributes.map(attributeName => {
+      // Object lookups will force any variable to strings, we don't want that for special objects etc
+      if (typeof attributeName !== 'string') {
+        return attributeName;
+      }
+
+      // Map attributes to column names
+      const columnName: string | undefined = Model.rawAttributes[attributeName]?.field;
+      if (columnName && columnName !== attributeName) {
+        return [columnName, attributeName];
+      }
+
+      return attributeName;
+    });
+  }
+
+  if (options.where != null && isPlainObject(options.where)) {
+    out.where = mapWhereFieldNames(options.where, Model);
+  }
+
+  return out;
+}
+
+export function mapWhereFieldNames(where: Record<string | symbol, any>, Model: ModelStatic<Model>): object {
+  if (!where) {
+    return where;
+  }
+
+  const newAttributes: Record<string | symbol, any> = Object.create(null);
+  // note on 'as any[]': TypeScript < 4.4 does not support using Symbol for keys.
+  // Cast can be removed in sept. 2022 when we drop support for < 4.4
+  for (const attributeNameOrOperator of getComplexKeys(where) as any[]) {
+    const rawAttribute: ModelAttributeColumnOptions | undefined = Model.rawAttributes[attributeNameOrOperator];
+
+    // note on 'any': TypeScript < 4.4 does not support using Symbol for keys.
+    // Cast can changed back to 'symbol | string' in sept. 2022 when we drop support for < 4.4
+    const columnNameOrOperator: any = rawAttribute?.field ?? attributeNameOrOperator;
+
+    if (
+      isPlainObject(where[attributeNameOrOperator])
+        && !(
+          rawAttribute
+          && (rawAttribute.type instanceof DataTypes.HSTORE
+            || rawAttribute.type instanceof DataTypes.JSON)
+        )
+    ) {
+      // Prevent renaming of HSTORE & JSON fields
+      newAttributes[columnNameOrOperator] = mapOptionFieldNames(
+        {
+          where: where[attributeNameOrOperator],
+        },
+        Model,
+      ).where;
+
+      continue;
+    }
+
+    if (Array.isArray(where[attributeNameOrOperator])) {
+      newAttributes[attributeNameOrOperator] = [...where[attributeNameOrOperator]];
+
+      for (const [index, wherePart] of where[attributeNameOrOperator].entries()) {
+        if (isPlainObject(wherePart)) {
+          newAttributes[columnNameOrOperator][index] = mapWhereFieldNames(wherePart, Model);
+        }
+      }
+
+      continue;
+    }
+
+    newAttributes[columnNameOrOperator] = where[attributeNameOrOperator];
+  }
+
+  return newAttributes;
+}
+
+/**
+ * getComplexKeys
+ *
+ * @param  {object} obj
+ * @returns {Array<string|symbol>} All keys including operators
+ * @private
```

Comment:
Not really specific to this line, but what was the consensus reached on marking functions as private? Or is this just for the docs generation?

Reply:
I think it's for docs generation

Reply:
I was wondering in general if we can remove the bigger parts of those comments since input and output types should be retrievable from the function signature?

Reply:
I expected eslint to at least remove types automatically but I forgot to turn `jsdoc/no-types` on (https://github.com/sequelize/sequelize/pull/13930#issuecomment-1013703583)

Reply:
👌